File: cpp/34. find-first-and-last-position-of-element-in-sorted-array.cpp
//
// Created by Rutvik Patel on 1/25/24.
//

#include "include.h"

// solution

class Solution {
private:
    std::vector<int> expandRange(std::vector<int>& nums, int target, int h_idx) {
        int s_idx, e_idx;

        for (s_idx = h_idx; s_idx >= 0; --s_idx)
            if (nums[s_idx] != target) break;

        for (e_idx = h_idx; e_idx < nums.size(); ++e_idx)
            if (nums[e_idx] != target) break;

        return { ++s_idx, --e_idx };
    }
    int binarySearch(std::vector<int>& nums, int target, int s_idx, int e_idx) {
        if (s_idx > e_idx) return -1;

        int m_idx = (s_idx + e_idx) / 2;

        if (target == nums[m_idx]) return m_idx;
        if (target < nums[m_idx]) return binarySearch(nums, target, s_idx, m_idx - 1);
        return binarySearch(nums, target, m_idx + 1, e_idx);
    }
public:
    std::vector<int> searchRange(std::vector<int>& nums, int target) {
        int h_idx = binarySearch(nums, target, 0, nums.size() - 1);
        if (h_idx == -1) return { -1, -1 };
        log ("h_idx: ", h_idx);
        return expandRange(nums, target, h_idx);
    }
};

// main

int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::pair<std::vector<int>, int>, std::vector<int>>> test_cases = {
        { { { 5, 7, 7, 8, 8, 10 }, 8 }, { 3, 4 } },
        { { { 5, 7, 7, 8, 8, 10 }, 6 }, { -1, -1 } },
        { { {  }, 0 }, { -1, -1 } },
        { { { 1 }, 1 }, { 0, 0 } },
    };

    Solution solution;

    for (auto& [args, ans]: test_cases)
        assert_m(solution.searchRange(args.first, args.second) == ans, "Failed!");

    log("All testcases passed!");
}

File: python/34. find-first-and-last-position-of-element-in-sorted-array.py
from typing import List


# lc sol start
class Solution:
    def __init__(self):
        pass

    def _binary_search(self, nums: List[int], target: int, s_idx: int, e_idx: int) -> int:
        if s_idx > e_idx:
            return -1

        m_idx = (s_idx + e_idx) // 2

        if target == nums[m_idx]:
            return m_idx

        if target < nums[m_idx]:
            return self._binary_search(nums, target, s_idx, m_idx - 1)

        return self._binary_search(nums, target, m_idx + 1, e_idx)

    def _get_hit_range(self, nums: List[int], target: int, h_idx: int):
        s_idx, e_idx = h_idx, h_idx

        while s_idx >= 0 and nums[s_idx] == target:
            s_idx -= 1

        while e_idx < len(nums) and nums[e_idx] == target:
            e_idx += 1

        return [s_idx + 1, e_idx - 1]

    def searchRange(self, nums: List[int], target: int) -> List[int]:
        h_idx = self._binary_search(nums, target, 0, len(nums) - 1)
        if h_idx == -1:
            return [-1, -1]
        return self._get_hit_range(nums, target, h_idx)

# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [([5, 7, 7, 8, 8, 10], 8), ([5, 7, 7, 8, 8, 10], 6), ([], 0), ([1], 1)],
        'outputs': [[3, 4], [-1, -1], [-1, -1], [0, 0]],
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        op = solution.searchRange(*args)
        assert op == test_cases['outputs'][idx], \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {op}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()
